tests/NTest/NTest.lua

```
@@ -0,0 +1,303 @@
+        if err then
+          err = err:match(".-([^\\/]*)$") -- cut off path of filename
+          if not err:match('_*_TestAbort_*_') then
+            handler('except', name, err)
+          end
+        end
+        if node then node.setonerror() end
+        copyenv(env, prev)
+        outputhandler('end', name)
+        table.remove(pendingtests, 1)
+        collectgarbage()
+        if next then next() end
+      end
+
+      local function wrap(method, ...)
+        method(handler, name, ...)
+      end
+
+      local function cbError(err)
+        err = err:match(".-([^\\/]*)$") -- cut off path of filename
+        if not err:match('_*_TestAbort_*_') then
+          handler('except', name, err)
+        end
+        restore()
+      end
+
+      env.eq = deepeq
+      env.spy = spy
+      env.ok = function (cond, msg1, msg2) wrap(assertok, false, cond, msg1, msg2) end
+      env.nok = function(cond, msg1, msg2) wrap(assertok, true,  cond, msg1, msg2) end
+      env.fail = function (func, expected, msg) wrap(fail, func, expected, msg) end
+
+      handler('begin', name)
+      node.setonerror(cbError)
+      local ok, err = pcall(f, async and restore)
+      if not ok then
+        err = err:match(".-([^\\/]*)$") -- cut off path of filename
+        if not err:match('_*_TestAbort_*_') then
+          handler('except', name, err)
+        end
+        if async then
+          restore()
+        end
+      end
+
+      if not async then
+        restore()
+      end
+    end
+
+    if not started then
+      outputhandler('start', testrunname)
+      started = true
+    end
+
+
+    table.insert(pendingtests, testfn)
+    if #pendingtests == 1 then
+      runpending()
+    end
+  end
+
+  local function test(name, f)
+    testimpl(name, f)
+  end
+
+  local function testasync(name, f)
+    testimpl(name, f, true)
+  end
+
+  local function report(f, envP)
+    if type(f) == 'function' then
+      outputhandler = f
+      env = envP or _G
+    end
+  end
```

Extremely minor and not worth not-approving the PR for: maybe rule out some dubious arguments rather than silently do nothing?
```suggestion
  local function report(f, envP)
    assert (f == nil or type(f) == 'function')
    assert (envP == nil or type(envP) == 'table')
    assert (f =~ nil or envP =~ nil)
    outputhandler = f or outputhandler
    env = envP or env
  end
```

---

Good point, the check was a leftover from when the interface was only one function to activate the special functionality.
As the other methods don't have any param checks I removed them here too and just left the selective update in place.